daemon/commands.c

```diff
@@ -0,0 +1,539 @@
+        cmd_unlock_high_priority
+};
+
+static cmd_status_t cmd_help_execute(char *args, char **message)
+{
+    (void)args;
+
+    *message = mallocz(MAX_COMMAND_LENGTH);
+    strncpyz(*message,
+             "\nThe commands are (arguments are in brackets):\n"
+             "\thelp\n"
+             "\treload health\n"
+             "\tsave database\n"
+             "\tlog rotate\n"
+             "\texit\n"
+             "\tfatal",
+             MAX_COMMAND_LENGTH - 1);
+
+    return CMD_STATUS_SUCCESS;
+}
+
+static cmd_status_t cmd_reload_health_execute(char *args, char **message)
+{
+    (void)args;
+    (void)message;
+
+    error_log_limit_unlimited();
+    info("COMMAND: Reloading HEALTH configuration.");
+    health_reload();
+    error_log_limit_reset();
+
+    return CMD_STATUS_SUCCESS;
+}
+
+static cmd_status_t cmd_save_database_execute(char *args, char **message)
+{
+    (void)args;
+    (void)message;
+
+    error_log_limit_unlimited();
+    info("COMMAND: Saving databases.");
+    rrdhost_save_all();
+    info("COMMAND: Databases saved.");
+    error_log_limit_reset();
+
+    return CMD_STATUS_SUCCESS;
+}
+
+static cmd_status_t cmd_log_rotate_execute(char *args, char **message)
+{
+    (void)args;
+    (void)message;
+
+    error_log_limit_unlimited();
+    info("COMMAND: Reopening all log files.");
+    reopen_all_log_files();
+    error_log_limit_reset();
+
+    return CMD_STATUS_SUCCESS;
+}
+
+static cmd_status_t cmd_exit_execute(char *args, char **message)
+{
+    (void)args;
+    (void)message;
+
+    error_log_limit_unlimited();
+    info("COMMAND: Cleaning up to exit.");
+    netdata_cleanup_and_exit(0);
+    exit(0);
+
+    return CMD_STATUS_SUCCESS;
+}
+
+static cmd_status_t cmd_fatal_execute(char *args, char **message)
+{
+    (void)args;
+    (void)message;
+
+    fatal("COMMAND: netdata now exits.");
+
+    return CMD_STATUS_SUCCESS;
+}
+
+static void cmd_lock_exclusive(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_wrlock(&exclusive_rwlock);
+}
+
+static void cmd_lock_orthogonal(unsigned index)
+{
+    uv_rwlock_rdlock(&exclusive_rwlock);
+    uv_mutex_lock(&command_lock_array[index]);
+}
+
+static void cmd_lock_idempotent(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_rdlock(&exclusive_rwlock);
+}
+
+static void cmd_lock_high_priority(unsigned index)
+{
+    (void)index;
+}
+
+static void cmd_unlock_exclusive(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_wrunlock(&exclusive_rwlock);
+}
+
+static void cmd_unlock_orthogonal(unsigned index)
+{
+    uv_rwlock_rdunlock(&exclusive_rwlock);
+    uv_mutex_unlock(&command_lock_array[index]);
+}
+
+static void cmd_unlock_idempotent(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_rdunlock(&exclusive_rwlock);
+}
+
+static void cmd_unlock_high_priority(unsigned index)
+{
+    (void)index;
+}
+
+static void pipe_write_cb(uv_write_t* req, int status)
+{
+    (void)status;
+    uv_pipe_t *client = req->data;
+
+    uv_close((uv_handle_t *)client, NULL);
```

Comment:
This implies that the command server allows exactly one command per connection - but the positioning of the call-backs in the sequence, the mutexes and the dynamic allocation of memory inside `parse_commands` suggest otherwise.
Basic design questions:
* How many active connections are allowed?
* How many commands are allowed over each connection?

If the answers are "one" and "one" then the code is too complex and needs to be simplified. If the answers are different then it's unclear why the parsing is being triggered at the end of the pipe read and the connection closed after execution of a command.

Reply:
Active connections can be many.

The parsing is triggered at the end because each command sends an EOF. The connection is closed only after the status is sent back to `cli`, which will be after the command execution.